scripts/fetch.py

```diff
@@ -0,0 +1,156 @@
+
+
+def provide_extraction_dir():
+    if not os.path.exists(EXTRACTION_DIR):
+        os.makedirs(EXTRACTION_DIR, 0o755)
+
+
+def remove_extraction_dir():
+    shutil.rmtree(EXTRACTION_DIR, True)
+
+
+def determine_number_of_images(dlinks_folder):
+    with open(os.path.join(dataset_dir, dlinks_folder, 'dlinks.txt')) as \
+            dlinks_file:
+        return sum(1 for line in dlinks_file)
+
+
+def collect_dlinks_lines():
+    lines = []
+    for folder in DLINKS_FOLDERS:
+        filename = os.path.join(dataset_dir, folder, 'dlinks.txt')
+        image_dir = os.path.join(os.path.dirname(filename), IMAGE_SUBDIR)
+        if not os.path.exists(image_dir):
+            os.mkdir(image_dir, 0o755)
+
+        with open(filename) as \
+                dlinks_file:
+            lines.extend([[line.rstrip('\n'), folder] for line in dlinks_file])
+
+    return lines
+
+
+def download_and_extract_archive(archive_url, pmc_id, extraction_dir_name,
+                                 image_name,
+                                 target_filename):
+    # download archive if it doesn't exist
+    archive_filename = os.path.join(extraction_dir_name,
+                                    archive_url.split(os.sep)[-1])
+
+    if not os.path.exists(archive_filename):
+        subprocess.call(
+            ['wget', '-nd', '-q', '-P', extraction_dir_name, archive_url])
+
+    archive_tarfile = tarfile.open(archive_filename)
+
+    image_name_in_archive = pmc_id + os.sep + image_name
+    image_tarinfo = archive_tarfile.getmember(image_name_in_archive)
+
+    # extract image to extraction dir, then copy to target file name
+    archive_tarfile.extractall(extraction_dir_name, [image_tarinfo])
+    image_filename = os.path.join(extraction_dir_name, image_name_in_archive)
+    shutil.copy(image_filename, target_filename)
+
+    # remove image and archive from extraction dir
+    shutil.rmtree(os.path.join(extraction_dir_name, pmc_id), True)
+    if not KEEP_ARCHIVES:
+        os.remove(archive_filename)
+
+
+def process_line(line_and_folder):
+    image_dir = os.path.join(dataset_dir, line_and_folder[1], IMAGE_SUBDIR)
+    archive_url, image_name, pmc_id, target_filename \
+        = extract_image_info(line_and_folder[0], image_dir)
+    with counter.get_lock():
+        counter.value += 1
+        log_status(counter.value, target_filename, num_images)
+    if not os.path.exists(target_filename):
+        download_and_extract_archive(
+            archive_url, pmc_id, EXTRACTION_DIR,
+            image_name, os.path.join(image_dir, target_filename))
+
+
+counter = None
+
+if __name__ == '__main__':
+    print('\rFetching ROCO dataset images...')
+    dataset_dir = os.path.dirname(os.path.abspath(os.path.realpath(sys.argv[0])
+                                                  + '/..'))
+    lines = collect_dlinks_lines()
+    num_images = len(lines)
+    provide_extraction_dir()
+
+    counter = multiprocessing.Value('i', 0)
```

Comment:
Why do you use a inter-process counter? I don't see any other usage of the value than printing a log message. Counting processed items would be much more efficient in the parent process using `imap_unordered` (although it should not be a bottleneck, but it is still a lock which affects all processes). This would even allow to use a progressbar from the `progressbar2` package.

Reply:
I changed it to use a loop around `imap_unordered`. Decided to forego using `progressbar2` for now because it would add a dependency. Wouldn't mind adding it though if you prefer that.